packages/ifpack2/src/Ifpack2_ILUT_def.hpp

```diff
@@ -744,11 +951,96 @@ void ILUT<MatrixType>::compute ()
+      if(A_local_crs.is_null()) {
+        local_ordinal_type numRows = A_local_->getLocalNumRows();
+        Array<size_t> entriesPerRow(numRows);
+        for(local_ordinal_type i = 0; i < numRows; i++) {
+          entriesPerRow[i] = A_local_->getNumEntriesInLocalRow(i);
+        }
+        RCP<crs_matrix_type> A_local_crs_nc =
+          rcp (new crs_matrix_type (A_local_->getRowMap (),
+                                    A_local_->getColMap (),
+                                    entriesPerRow()));
+        // copy entries into A_local_crs
+        nonconst_local_inds_host_view_type indices("indices",A_local_->getLocalMaxNumRowEntries());
+        nonconst_values_host_view_type values("values",A_local_->getLocalMaxNumRowEntries());
+        for(local_ordinal_type i = 0; i < numRows; i++) {
+          size_t numEntries = 0;
+          A_local_->getLocalRowCopy(i, indices, values, numEntries);
+          A_local_crs_nc->insertLocalValues(i, numEntries, reinterpret_cast<scalar_type*>(values.data()),indices.data());
+        }
+        A_local_crs_nc->fillComplete (A_local_->getDomainMap (), A_local_->getRangeMap ());
+        A_local_crs = rcp_const_cast<const crs_matrix_type> (A_local_crs_nc);
+      }
+      auto lclMtx = A_local_crs->getLocalMatrixDevice();
+      A_local_rowmap_  = lclMtx.graph.row_map;
+      A_local_entries_ = lclMtx.graph.entries;
+      A_local_values_  = lclMtx.values;
+    }
+
+      //JHU TODO Should allocation of L & U's column (aka entry) and value arrays occur here or in init()?
+      auto par_ilut_handle = KernelHandle_->get_par_ilut_handle();
+      auto nnzL = par_ilut_handle->get_nnzL();
+      L_entries_ = static_graph_entries_t("L_entries", nnzL);
+      L_values_ = local_matrix_values_t("L_values", nnzL);
+
+      auto nnzU = par_ilut_handle->get_nnzU();
+      U_entries_ = static_graph_entries_t("U_entries", nnzU);
+      U_values_ = local_matrix_values_t("U_values", nnzU);
+
+      KokkosSparse::Experimental::par_ilut_numeric(KernelHandle_.getRawPtr(),
+                                                   A_local_rowmap_, A_local_entries_, A_local_values_,
+                                                   L_rowmap_, L_entries_, L_values_, U_rowmap_, U_entries_, U_values_);
+
+    auto L_kokkosCrsGraph = local_graph_device_type(L_entries_, L_rowmap_);
+    auto U_kokkosCrsGraph = local_graph_device_type(U_entries_, U_rowmap_);
+
+    using local_matrix_device_type = typename crs_matrix_type::local_matrix_device_type;
+    local_matrix_device_type L_localCrsMatrix_device;
+    L_localCrsMatrix_device = local_matrix_device_type("L_Factor_localmatrix",
+                                                       A_local_->getLocalNumRows(),
+                                                       L_values_,
+                                                       L_kokkosCrsGraph);
+
+    L_ = rcp (new crs_matrix_type (L_localCrsMatrix_device,
```

Comment:
Since ParILUT is a local-only preconditioner, is it possible to just use the local L and U directly for apply, instead of making Tpetra CrsMatrices?

Reply:
@brian-kelley This is because Ifpack2's LocalSparseTriangularSolver class takes a Tpetra::CrsMatrix.  Something similar is done in RILUK for the case of the spiluk algorithm.